File: typescript/src/data-structures/hash-table.ts
/**
 * Hash table implementation with open addressing and linear probing.
 * Time: O(1) average for all operations
 * Space: O(n)
 */

interface HashEntry<K, V> {
  key: K;
  value: V;
  deleted: boolean;
}

export class HashTable<K, V> {
  private buckets: (HashEntry<K, V> | null)[];
  private _size: number = 0;
  private _capacity: number;
  private readonly loadFactorThreshold = 0.7;

  constructor(initialCapacity: number = 16) {
    this._capacity = Math.max(1, initialCapacity);
    this.buckets = new Array(this._capacity).fill(null);
  }

  put(key: K, value: V): void {
    if (this._size >= this._capacity * this.loadFactorThreshold) {
      this.resize(this._capacity * 2);
    }

    let index = this.hash(key);
    let firstDeleted = -1;

    for (let i = 0; i < this._capacity; i++) {
      const entry = this.buckets[index];

      if (entry === null) {
        const insertIndex = firstDeleted !== -1 ? firstDeleted : index;
        this.buckets[insertIndex] = { key, value, deleted: false };
        this._size++;
        return;
      }

      if (entry.deleted && firstDeleted === -1) {
        firstDeleted = index;
      } else if (!entry.deleted && this.keysEqual(entry.key, key)) {
        entry.value = value;
        return;
      }

      index = (index + 1) % this._capacity;
    }

    if (firstDeleted !== -1) {
      this.buckets[firstDeleted] = { key, value, deleted: false };
      this._size++;
    }
  }

  get(key: K): V | undefined {
    const index = this.findIndex(key);
    if (index === -1) {
      return undefined;
    }
    return this.buckets[index]!.value;
  }

  remove(key: K): V | undefined {
    const index = this.findIndex(key);
    if (index === -1) {
      return undefined;
    }
    const entry = this.buckets[index]!;
    const value = entry.value;
    entry.deleted = true;
    this._size--;
    return value;
  }

  has(key: K): boolean {
    return this.findIndex(key) !== -1;
  }

  isEmpty(): boolean {
    return this._size === 0;
  }

  size(): number {
    return this._size;
  }

  clear(): void {
    this.buckets = new Array(this._capacity).fill(null);
    this._size = 0;
  }

  keys(): K[] {
    const result: K[] = [];
    for (const entry of this.buckets) {
      if (entry && !entry.deleted) {
        result.push(entry.key);
      }
    }
    return result;
  }

  values(): V[] {
    const result: V[] = [];
    for (const entry of this.buckets) {
      if (entry && !entry.deleted) {
        result.push(entry.value);
      }
    }
    return result;
  }

  private hash(key: K): number {
    const str = String(key);
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      hash = (hash * 31 + str.charCodeAt(i)) >>> 0;
    }
    return hash % this._capacity;
  }

  private keysEqual(a: K, b: K): boolean {
    return a === b;
  }

  private findIndex(key: K): number {
    let index = this.hash(key);

    for (let i = 0; i < this._capacity; i++) {
      const entry = this.buckets[index];

      if (entry === null) {
        return -1;
      }

      if (!entry.deleted && this.keysEqual(entry.key, key)) {
        return index;
      }

      index = (index + 1) % this._capacity;
    }

    return -1;
  }

  private resize(newCapacity: number): void {
    const oldBuckets = this.buckets;
    this._capacity = newCapacity;
    this.buckets = new Array(this._capacity).fill(null);
    this._size = 0;

    for (const entry of oldBuckets) {
      if (entry && !entry.deleted) {
        this.put(entry.key, entry.value);
      }
    }
  }
}

File: swift/Sources/DSA/DataStructures/HashTable.swift
/// Hash table implementation with open addressing and linear probing.
/// Time: O(1) average for all operations
/// Space: O(n)
public struct HashTable<K: Hashable, V> {
    private struct Entry {
        let key: K
        var value: V
        var deleted: Bool = false
    }

    private var buckets: [Entry?]
    private var size = 0
    private let loadFactor = 0.7

    public init(capacity: Int = 16) {
        buckets = Array(repeating: nil, count: max(1, capacity))
    }

    public mutating func put(_ key: K, _ value: V) {
        if Double(size) >= Double(buckets.count) * loadFactor {
            resize(buckets.count * 2)
        }

        var index = hash(key)
        var firstDeleted = -1

        for _ in 0..<buckets.count {
            if let entry = buckets[index] {
                if entry.deleted && firstDeleted == -1 {
                    firstDeleted = index
                } else if !entry.deleted && entry.key == key {
                    buckets[index]?.value = value
                    return
                }
            } else {
                let insertIndex = firstDeleted != -1 ? firstDeleted : index
                buckets[insertIndex] = Entry(key: key, value: value)
                size += 1
                return
            }
            index = (index + 1) % buckets.count
        }

        if firstDeleted != -1 {
            buckets[firstDeleted] = Entry(key: key, value: value)
            size += 1
        }
    }

    public func get(_ key: K) -> V? {
        if let index = findIndex(key) {
            return buckets[index]?.value
        }
        return nil
    }

    public mutating func remove(_ key: K) -> V? {
        if let index = findIndex(key) {
            let value = buckets[index]?.value
            buckets[index]?.deleted = true
            size -= 1
            return value
        }
        return nil
    }

    public func contains(_ key: K) -> Bool {
        findIndex(key) != nil
    }

    public var isEmpty: Bool {
        size == 0
    }

    public var count: Int {
        size
    }

    public func keys() -> [K] {
        buckets.compactMap { entry in
            guard let entry = entry, !entry.deleted else { return nil }
            return entry.key
        }
    }

    public func values() -> [V] {
        buckets.compactMap { entry in
            guard let entry = entry, !entry.deleted else { return nil }
            return entry.value
        }
    }

    public mutating func clear() {
        buckets = Array(repeating: nil, count: buckets.count)
        size = 0
    }

    private func hash(_ key: K) -> Int {
        abs(key.hashValue) % buckets.count
    }

    private func findIndex(_ key: K) -> Int? {
        var index = hash(key)

        for _ in 0..<buckets.count {
            guard let entry = buckets[index] else { return nil }
            if !entry.deleted && entry.key == key {
                return index
            }
            index = (index + 1) % buckets.count
        }

        return nil
    }

    private mutating func resize(_ newCapacity: Int) {
        let oldBuckets = buckets
        buckets = Array(repeating: nil, count: newCapacity)
        size = 0

        for entry in oldBuckets {
            if let entry = entry, !entry.deleted {
                put(entry.key, entry.value)
            }
        }
    }
}
